pkg/yurthub/proxy/autonomy/autonomy.go

```diff
@@ -0,0 +1,210 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package autonomy
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"sync/atomic"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/pointer"
+
+	appsv1beta1 "github.com/openyurtio/openyurt/pkg/apis/apps/v1beta1"
+	"github.com/openyurtio/openyurt/pkg/projectinfo"
+	"github.com/openyurtio/openyurt/pkg/yurthub/cachemanager"
+	hubrest "github.com/openyurtio/openyurt/pkg/yurthub/kubernetes/rest"
+	proxyutil "github.com/openyurtio/openyurt/pkg/yurthub/proxy/util"
+	"github.com/openyurtio/openyurt/pkg/yurthub/util"
+)
+
+var (
+	nodeStatusUpdateRetry = 5
+	ErrRestConfigMgr      = fmt.Errorf("failed to initialize restConfigMgr")
+)
+
+type AutonomyProxy struct {
+	cacheMgr         cachemanager.CacheManager
+	restConfigMgr    *hubrest.RestConfigManager
+	cacheFailedCount *int32
+}
+
+func NewAutonomyProxy(
+	restConfigMgr *hubrest.RestConfigManager,
+	cacheMgr cachemanager.CacheManager,
+) *AutonomyProxy {
+	return &AutonomyProxy{
+		restConfigMgr:    restConfigMgr,
+		cacheMgr:         cacheMgr,
+		cacheFailedCount: pointer.Int32(0),
+	}
+}
+
+func (ap *AutonomyProxy) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	obj, err := ap.updateNodeStatus(req)
+	if err != nil {
+		proxyutil.Err(err, rw, req)
+	}
+	util.WriteObject(http.StatusOK, obj, rw, req)
+}
+
+func (ap *AutonomyProxy) updateNodeStatus(req *http.Request) (runtime.Object, error) {
+	_, ok := apirequest.RequestInfoFrom(req.Context())
+	if !ok {
+		return nil, fmt.Errorf("failed to resolve request")
+	}
+
+	var node runtime.Object
+	var err error
+	for i := 0; i < nodeStatusUpdateRetry; i++ {
+		node, err = ap.tryUpdateNodeConditions(i, node, req)
+		if errors.Is(err, ErrRestConfigMgr) {
+			break
+		} else if err != nil {
+			klog.ErrorS(err, "Error getting or updating node status, will retry")
+		} else {
+			return node, nil
+		}
+	}
+	if node == nil {
+		return nil, fmt.Errorf("failed to get node")
+	}
+	klog.ErrorS(err, "failed to update node autonomy status")
+	return node, nil
+}
+
+func (ap *AutonomyProxy) tryUpdateNodeConditions(tryNumber int, node runtime.Object, req *http.Request) (runtime.Object, error) {
+	var originalNode, updatedNode *v1.Node
+	var err error
+	info, _ := apirequest.RequestInfoFrom(req.Context())
+
+	if node == nil {
+		if tryNumber == 0 {
+			// get from local cache
+			obj, err := ap.cacheMgr.QueryCache(req)
+			if err != nil {
+				return nil, err
+			}
+			ok := false
+			originalNode, ok = obj.(*v1.Node)
+			if !ok {
+				return nil, fmt.Errorf("could not QueryCache, node is not found")
+			}
+		} else {
+			if ap.restConfigMgr == nil {
+				return nil, fmt.Errorf("failed to initialize restConfigMgr")
```

Comment:
fmt.Errorf("failed to initialize restConfigMgr") --> ErrRestConfigMgr

Reply:
done